### DISCUSSION TITLE
Chat GPT integration

### DISCUSSION THREAD

**Discussion by dpolivaev** [freeplane/freeplane#1048] [Ideas]
I am looking for ways to integrate Chat GPT. In the beginning, I let it generate some mind maps on given topics in the classical mind map format and in the trigger words format proposed by my friend.

I am sharing my results here.

[mindMapsByAI.zip](https://github.com/freeplane/freeplane/files/11008593/mindMapsByAI.zip)

Obviously I am not start working on it until Freeplane 1.11 is released

**Comment by euu2021** [comment]
I don't know what is possible to do with this integration, but It would be awesome if the user could ask ChatGPT to perform actions in FP, like an AI personal assistant: 

- Inside the branch of the selected node, find nodes that have the style "ToImprove"; rewrite the content of the node core (in order to make it more clear for publishing as a scientific article); and set the node to the style "ImprovedByIA"

This is not something that can be done by copy and pasting text in the ChatGPT website. 

If it is possible to make the AI interact with the map like this, the possibilities that it opens are... interesting, to say the least 🥲

**Reply by dpolivaev** [reply]
Well, ChatGPT has an interface to send text and receive text. It knows nothing about Freeplane.
The integration can send requests containing text from e.g. the selected node and some related nodes, ask questions, receive an answer and use it as new content for one or more nodes. For instance, it could also set the style to ImprovedByAI or GeneratedByAI.

The attached maps show how the AI can generate content of whole mind maps or their substantial parts given just a single prompt. This content could be automatically added to the current map or to a new map.

**Reply by euu2021** [reply]
I see. I'm thinking about a structure where ChatGPT works as a script builder and the FP integration gives it the necessary information, when requested. 

But, I don't know if ChatGPT is smart enough to build FP scripts (I'm probably overestimating it, because of the anecdotal stories that I've heard 🤣). My knowledge about ChatGPT is very limited, but I've seen that the user can train it before it outputs the result ([this is the example ](https://medium.com/@JarrettYe/casting-a-spell-on-chatgpt-let-it-write-anki-cards-for-you-a-prompt-engineering-case-fd7d577b9d94)that I've seemed). So, maybe, it's possible to teach it about the FP API before asking it to create the script.

Something that goes like this:

User asks FP: 

- Inside the branch of the selected node, find nodes that have the style "ToImprove"; rewrite the content of the node core (in order to make it more clear for publishing as a scientific article); and set the node to the style "ImprovedByIA"

FP sends to ChatGPT

- This is how to use the FP API...
- Now, creates a script that does... (what the user asked)
- If, necessary, you can ask me for information about the map. Ask in this format...

ChatGPT to FP

- What is the ID number of the currently selected node?

FP to ChatGPT

- 239482389

ChatGPT

- Here is the script...

FP grabs the script and applies it (maybe do some kind of validation and debugging before applying)

**Reply by dpolivaev** [reply]
Actually, I am rather interested in using ChatGPT as a content assistant.

**Reply by euu2021** [reply]
Of course, I was just dreaming awake. AI is not ready to do those things, yet. But, it's exciting to think that soon it will be.

**Reply by GorBinance** [reply]
Greetings dear friends, it makes sense to integrate with local open source code Alpaca, to use with a mind map as a personal assistant who knows you. I use the mind map to parse personalities and it's great! I'd be happy to assist.

**Reply by GorBinance** [reply]
Obviously a bit off-topic, but a question, is there any way to visualize pie charts with points from nodes in Freeplane?

If not, is there any idea to implement something similar in the plugin structure.

Creating a GRPC server that processes client requests and creates pie charts based on the received data.

Creating a GRPC client that contacts the GRPC server to create pie charts.

Creation of the Freeplane plugin that uses the GRPC client to create pie charts and render them into a mindmap.

An example of the plugin's algorithm:

The user opens the mindmap and selects the item for which they want to create a pie chart.

The user launches the Freeplane plugin, which launches a GRPC request to the server to get the data for the pie chart.

The GRPC server processes the request and creates a pie chart based on the data received.

The GRPC server sends the created pie chart back to the client.

The GRPC client receives the pie chart from the server and passes it to the Freeplane plugin.

The Freeplane plugin visualizes the pie chart as a mind map.

**Reply by dpolivaev** [reply]
Hello, @GorBinance if you are interested in contributing to the Freeplane development and integration of some tools, I suggest we discuss it in a video call.  Yes, we can have a call in Russian. But I appreciate that you use English here.

**Reply by GorBinance** [reply]
> Hello, @GorBinance if you are interested in contributing to the Freeplane development and integration of some tools, I suggest we discuss it in a video call. Yes, we can have a call in Russian. But I appreciate that you use English here.

That's nice to hear, how can I contact you?

**Reply by dpolivaev** [reply]
email: dimitry (at) freeplane (dot) org

**Comment by sifran-github** [comment]
Can you share the prompts that made chatgpt create these maps?

**Reply by dpolivaev** [reply]
To be used on https://platform.openai.com/playground?mode=chat

I entered the following prompt as a "system message". Then you can choose any topic you like as a "user message". The output can be copied and pasted into Freeplane. It is still a work in progress.

````
Learn the format "trigger word outline" from the example below and create similar trigger word outlines for topics given by the user. Strictly follow the format. Use multiple levels. Always break sentences. Put each word on its own line. Use no '-' characters or bullet points. Cover as many aspects as possible as deeply as possible. 

Example:

World War II
    Time Line
        Starts at
            1939
                September
                    Nazi Invade
                        Poland
            1931
                Japan
                    Invade
                        Manchuria
            1937
                Japan
                    Invade
                        China
                            They Did't stopped
                                Until 1945
            1933
                Germany
                    Hitler
                        Took Power
            1941
                US
                    Started
                        Fighting
    Where / When
        China
            Brutal Fighting
                In China
                    The Fighting was
                        Very Brutal
            Example
                Nanking Rape
                    Massive Slaughter
                        Where
                            Hundreds of Thousands
                                Were Slaughter
                    Still Controversial Today
                        Affects Relations
                            Between
                                Japan
                                China
        European Theater
            General Place for WW2
                Is the World War II
                    We mostly see on
                        Movies
                        Books
                        TV
````

**Reply by sifran-github** [reply]
i don't like this style of mind map. for example:
```
            1937
                Japan
                    Invade
                        China
                            They Did't stopped
                                Until 1945
```
what is this? its just a sentence broken into an outline style. its not even meaningful.

**Reply by dpolivaev** [reply]
Unfortunately, this kind of mind map is the only one I could get access to because it was created by my friend.
When I ask users to share their mind maps with me, they never respond.
So I have no other mind maps I could use in a prompt.

I would appreciate it if some other contentful examples could be made available for me.

**Reply by dpolivaev** [reply]
Here you can see how it looks
<img width="1207" alt="WWII-timeline" src="https://user-images.githubusercontent.com/1558052/229086372-91423ede-4f43-4855-aeda-36041821f22b.png">

**Reply by sifran-github** [reply]
well, you can simply go to https://chat.openai.com/chat (free) and use the following phrase: "write an outline that summarize the topic of: X" with X being your topic. it will give you an outline of the topic.

for example, "write an outline that summarize the topic of: Freeplane". the result:
```
I. Introduction
A. Definition of Freeplane
B. Brief history and development
C. Purpose of Freeplane

II. Features
A. Mind-mapping capabilities
B. Task management tools
C. Note-taking and organization
D. Customizable interface

III. How to Use Freeplane
A. Downloading and installing
B. Creating a mind map
C. Customizing the interface
D. Using task management tools

IV. Advantages and Disadvantages of Freeplane
A. Advantages
1. User-friendly interface
2. Wide range of features
3. Cross-platform compatibility
4. Open-source software
B. Disadvantages
1. Steep learning curve
2. Limited collaboration features

V. Conclusion
A. Summary of Freeplane
B. Final thoughts
C. Future developments and updates.
```

I find this outline "nice to have", but not so amazing.
If you want to create a course on some topic, or you know nothing about some topic and you want to break it down, it can be nice.

**Reply by dpolivaev** [reply]
If you prefer different styles and don't like the idea of sharing your content you could rework this mind map so that I can see what styles you like more:

[Freeplane Mind Map - World War II.mm.zip](https://github.com/freeplane/freeplane/files/11120572/Freeplane.Mind.Map.-.World.War.II.mm.zip)

**Reply by sifran-github** [reply]
> Here you can see how it looks <img alt="WWII-timeline" width="1207" src="https://user-images.githubusercontent.com/1558052/229086372-91423ede-4f43-4855-aeda-36041821f22b.png">

I find this map to be confusing and not useful at all.
simple example: September at the same level as Japan and other countries? (the blue columns). September is a month, not a country.
I can find many other weird things in this mind map, and i will never use it as it is.

**Reply by dpolivaev** [reply]
If you find anything weird, it is your opinion which may be agreed by some and disagreed by others.
I find making such claims not helpful.
Instead, I suggest that you share an example that you personally like. Then I can use it in a prompt too and give users a choice.

**Reply by sifran-github** [reply]
ok but what is the purpose that you are trying to achieve?

to share nice mind map structures?
to understand how to get chatgpt to write useful maps?
other goal?

**Reply by dpolivaev** [reply]
> well, you can simply go to https://chat.openai.com/chat (free) and use the following phrase: "write an outline that summarize the topic of: X" with X being your topic. it will give you an outline of the topic.
> 
> for example, "write an outline that summarize the topic of: Freeplane". the result:

I find Knowledge of ChatGPT 3.5 about Freeplane pretty limited, and therefore it is not a good example.
Also, its understanding of outlines usually does not go much in depth unless I give an example in the prompt.

**Reply by dpolivaev** [reply]
> ok but what is the purpose that you are trying to achieve?
> 
> to share nice mind map structures? to understand how to get chatgpt to write useful maps? other goal?

I want to use nice mind map outline examples in a prompt as I have shown so that  ChatGPT can be used to generate a mind map on any common knowledge topic or structure or summarize any user information as a mind map. I want to integrate it in Freeplane so that any node or part of the mind map can be refined or restructured using ChatGPT.

**Reply by sifran-github** [reply]
I'm having hard time now to think about how exactly I want a mind-map output to look like. The reason is that for a computer, the same information can be represented in many ways: in a mind map, in an outline, in a document with titles and subtitles and more formats.

In my opinion there are advantages to a mind map format over other formats, for example the ability to put a lot of the information in notes, therefore having a clean "big picture" without all the small details, and it helps to understand, think and rearrange the content better.

but again, if we are talking about some AI that generates it, i just wonder whats even the difference between a mind map output, and for example... the structure of articles in Wikipedia. You can take every title of an article as a node text, and the full text in notes...

I don't know, I'll have to think about it.
maybe someone can define it better, and i will be able to contribute more.

**Comment by dpolivaev** [comment]
Has someone experimented with ChatGPT prompts? Do you have any prompts or maps to share?

**Reply by sifran-github** [reply]
Surprisingly, I don't use chatgpt to create outlines very often.
I'm not sure why. Probably its the quality of chatgpt answers, specially when i try to instruct it to write an outline.
I'm using the free chatgpt 3.5

**Reply by dpolivaev** [reply]
I am using the free version too. And given some good examples showing the structure and the depth using some particular topic I can get it to create a similar output for any other common knowledge topic which can be further refined.

**Reply by dpolivaev** [reply]
By the way, to create a formatted outline with proper indentation I use https://platform.openai.com/playground?mode=chat and not the usual chat interface.

**Comment by barrymac** [comment]
I really like this initiative. I've tried generating freeplane XML with GPT and it didn't do a great job, and none of them could be opened. But I'm thinking that maybe a groovy plugin with the GPT API could work pretty well and could even be an interesting way to do things like what is being done with AutoGPT/BabyAGI/Agent-LLM and the like. this is a nice demo of how this can work:  https://www.superusapp.com/map

**Reply by barrymac** [reply]
Well, I started an addon. It's very basic but it's already quite fun to use I think. https://github.com/barrymac/freeplane_openai_addon

**Reply by ldmpub** [reply]
Video of the OpenAI Freeplane add-on working
![demo](https://user-images.githubusercontent.com/1452146/237060693-22af861a-3da9-460f-b057-12820a62eeeb.gif)

Source of inspiration (quite the same video but not in Freeplane 😋) : https://www.superusapp.com

**Comment by dpolivaev** [comment]
Hello, @barrymac, and all other interested people.

I want to share with you two prompts I used as a system message to generate a text outline that can be parsed by Freeplane. I am sorry for the example topic choice. I just used what I got.

To create a branch from this topic you currently need to copy it into the system clipboard as text and trigger Edit->Paste menu action.

I was going to use it in my ChatGPT integration and also to extend scripting API so that "paste" can be called directly on a node. But I want to help you to experiment with the prompts right now, so I am sharing what I have. I used the promts as system messages with completion mode  `chat` with model `gpt-3.5-turbo` and maximum response length 1471.
I am also attaching example outputs for User message `South Africa` added to the each system prompt.

### Prompt version 1: world war II prompt
System message:
````
This is an outline about World war II

World War II
	Introduction
		World War II was a global conflict that lasted from 1939 to 1945. It was the deadliest conflict in human history, with an estimated 50 million to 85 million fatalities.
	Causes
		The main causes of World War II were the ambitions of totalitarian powers in Germany, Italy, and Japan, and their aggression towards other nations.
		Other contributing factors included the Treaty of Versailles, which punished Germany after World War I, and the failure of appeasement policies in dealing with Nazi Germany's territorial demands.
	Major Events
		1939 - Germany invades Poland, starting the war in Europe.
		1940 - Germany invades Denmark, Norway, Belgium, the Netherlands, and France.
		1941 - Japan attacks Pearl Harbor, bringing the United States into the war.
		1942 - Allies win key battles in North Africa and the Pacific.
		1943 - Allies invade Italy and begin to push back German forces in the Soviet Union.
		1944 - Allies invade France on D-Day, leading to the liberation of Paris.
		1945 - Allied forces capture Berlin and Germany surrenders; the United States drops atomic bombs on Hiroshima and Nagasaki, leading to Japan's surrender.
	Major Powers
		Axis Powers:
			Germany, Italy, and Japan
		Allied Powers:
			United States, Great Britain, Soviet Union, and China
	Impact
		The war had a profound impact on the world, leading to the formation of the United Nations and the emergence of the United States and Soviet Union as superpowers.
		Over 70 million military personnel were mobilized, making it the largest and most widespread war in history.
		Countless civilians were also affected, with many being killed, injured, or displaced.
		The war also led to the Holocaust, in which millions of Jews and other minority groups were systematically murdered by Nazi Germany.
	Legacy
		World War II had a lasting impact on the world, shaping the political, social, and economic landscape for decades to come.
		It also led to the development of new technologies and innovations, including the atomic bomb, and marked a turning point in global affairs.
		
Create a similar outline using nested outline levels to break down the content about the topics suggested by the user. Start a new paragraph for each new sentence. Suppress any bullet points or other characters like - at the beginning of the list, use just the formatting as in the example 
````

Example output:
````
South Africa
	Introduction
		South Africa is a country located at the southern tip of the African continent.
		It has a diverse population, with a rich cultural heritage and a history marked by colonialism and apartheid.
	History
		South Africa was colonized by the Dutch and later the British, leading to centuries of racial discrimination and oppression.
		Apartheid, a system of segregation and discrimination against non-white South Africans, was introduced in 1948 and lasted until 1994.
		Nelson Mandela, a prominent anti-apartheid activist, became the country's first black president in 1994 after years of struggle and resistance.
	Politics
		South Africa is a constitutional democracy, with a president as the head of state and a multi-party system.
		The African National Congress (ANC) has been the ruling party since the end of apartheid.
		The country faces significant challenges, including high levels of poverty, inequality, and corruption.
	Economy
		South Africa has the second-largest economy in Africa after Nigeria.
		It is rich in natural resources, including gold, diamonds, and platinum.
		However, the country faces significant economic challenges, including high unemployment and inequality.
	Culture
		South Africa has a rich cultural heritage, with diverse traditions and languages.
		Music, dance, and art are an integral part of the country's cultural identity.
		South Africa is also known for its wildlife, including the "Big Five" animals: lions, leopards, elephants, rhinoceroses, and buffaloes.
	Tourism
		South Africa is a popular tourist destination, known for its natural beauty, wildlife, and cultural attractions.
		Visitors can enjoy safaris, wineries, beaches, and historical landmarks such as Robben Island, where Nelson Mandela was imprisoned for many years.
		Tourism plays a significant role in the country's economy and provides employment opportunities for many South Africans. 
````

### Prompt version 2: world war II trigger words prompt

System message:
````
Learn the format "trigger word outline" from the example below and create similar trigger word outlines for topics given by the user. Strictly follow the format. Use multiple nested levels. Always break sentences. Put each word on its own line. Use no '-' characters or bullet points. Cover as many aspects as possible as deeply as possible. 

Example:

World War II
    Time Line
        Starts at
            1939
                September
                    Nazi Invade
                        Poland
            1931
                Japan
                    Invade
                        Manchuria
            1937
                Japan
                    Invade
                        China
                            They Did't stopped
                                Until 1945
            1933
                Germany
                    Hitler
                        Took Power
            1941
                US
                    Started
                        Fighting
    Where / When
        China
            Brutal Fighting
                In China
                    The Fighting was
                        Very Brutal
            Example
                Nanking Rape
                    Massive Slaughter
                        Where
                            Hundreds of Thousands
                                Were Slaughter
                    Still Controversial Today
                        Affects Relations
                            Between
                                Japan
                                China
        European Theater
            General Place for WW2
                Is the World War II
                    We mostly see on
                        Movies
                        Books
                        TV
````

Example output:
````
South Africa
    History
        Apartheid Era
            Started in
                1948
            Separated People
                By Race
            Ended in
                1994
            Nelson Mandela
                First Black President
    Geography
        Location
            Southernmost Country in Africa
        Borders
            Namibia
            Botswana
            Zimbabwe
            Mozambique
            Eswatini
            Lesotho
        Natural Wonders
            Table Mountain
            Kruger National Park
            Cape of Good Hope
    Culture
        Languages
            11 Official Languages
                Including
                    Zulu
                    Xhosa
                    Afrikaans
                    English
        Cuisine
            Braai (Barbecue)
            Biltong (Dried Meat)
            Bunny Chow (Curry Dish)
        Music
            Traditional
                Mbube
                Maskandi
            Modern
                Kwaito
                Gqom
    Politics
        Government
            Constitutional Democracy
        Current Issues
            Corruption
            High Crime Rate
            Land Reform and Redistribution
    Economy
        Major Industries
            Mining
            Agriculture
            Tourism
        Challenges
            High Unemployment Rate
            Income Inequality
            Dependence on Natural Resources
````

**Comment by Illegit** [comment]
If using the addon, how to ask a chain of questions instead than only one?

**Reply by dpolivaev** [reply]
The chain of questions is much harder to implement because of the limited number of tokens ChatGPT can process. Our current efforts are instead put in prompts generation.

**Comment by Illegit** [comment]
Ok, thankyou.

**Comment by Illegit** [comment]
Hi, I am asking for any suggestions because this addon is pure chatgpt openai connected, but what do we need to do if we have another url account (eg.: Bing with chatgpt).
1) how is possible to use the AskGPT addon with Bing? 
2) Is it only a matter of providing an appropriate URL?
3) If we have a Bing account, how we may be able to activate the map connection?

**Comment by fredlim** [comment]
Would you add local LLM to this add-on, I want to use Ollama or LM Studio without connecting to Open-ai.
